### PR TITLE
Improvements to external dependency management (CVM-1161)

### PR DESCRIPTION
**Pull request by radupopescu**
This branch introduces a series of improvements to the scripts that build external dependencies and to the CMake configuration files that handle the externals.

Summary:
* `bootstrap.sh` has be modified. It's called at the beginning of the CMake configuration process, as before, but it will now configure and build all the external dependencies at once. Building them is no longer delayed until the CVMFS build phase. This makes it simpler to understand the external build process
* After being built, the external libraries are all installed under the same prefix, by default `${PROJECT_SOURCE_DIR}/externals_install`. This means that they can easily be used by multiple builds of CVMFS, if desired. The location of the external libs can be set using the `EXTERNALS_PREFIX` CMake variable.
* It's possible to disable using built-in external libraries by giving the `BUILTIN_EXTERNALS=OFF` parameter to CMake.
* At the end of the external build process, the file `$EXTERNALS_INSTALL_LOCATION/.bootstrapDone` contains a list of all the externals that are present. By default, the `bootstrap.sh` script will not recompile any library listed here. To force the recompilation of any of them, simply remove their name from the file and re-run the CMake configuration for CVMFS.
* Various simplifications of the CMake configurations scrips:
    * When `BUILTIN_EXTERNALS=ON`, all the externals libraries are picked up from `EXTERNALS_INSTALL_LOCATION`. When `BUILTIN_EXTERNALS=OFF`, they are picked up from the default system prefix. In both cases, it's now possible to rely on the `find_package` standard mechanism.
    * Adding new Find****.cmake modules for external libs, where this was missing.
    * Cosmetic, cleanup etc.
* It's now possible to use the Ninja build system instead of Make. Just configure CVMFS with `cmake -G Ninja ../"

All these changes should not modify the output (executables, libraries etc.) of the build system in any way!

### CODE REVIEW

File: cmake/Modules/FindGOOGLEBENCH.cmake [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+    GOOGLEBENCH
+    DEFAULT_MSG
+    GOOGLEBENCH_LIBRARIES
+    GOOGLEBENCH_INCLUDE_DIRS
+)
+
+if(GOOGLEBENCH_FOUND)
+    mark_as_advanced(GOOGLEBENCH_LIBRARIES GOOGLEBENCH_INCLUDE_DIRS)
+endif()
```
Comment (jblomer) [inline cmake/Modules/FindGOOGLEBENCH.cmake]: A missing new line, but probably harmless

File: cmake/Modules/FindSHA2.cmake [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+    SHA2
+    DEFAULT_MSG
+    SHA2_LIBRARIES
+    SHA2_INCLUDE_DIRS
+)
+
+if(SHA2_FOUND)
+    mark_as_advanced(SHA2_LIBRARIES SHA2_INCLUDE_DIRS)
+endif()
```
Comment (jblomer) [inline cmake/Modules/FindSHA2.cmake]: Missing new line

File: cmake/Modules/FindSHA3.cmake [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+    SHA3
+    DEFAULT_MSG
+    SHA3_LIBRARIES
+    SHA3_INCLUDE_DIRS
+)
+
+if(SHA3_FOUND)
+    mark_as_advanced(SHA3_LIBRARIES SHA3_INCLUDE_DIRS)
+endif()
```
Comment (jblomer) [inline cmake/Modules/FindSHA3.cmake]: Missing newline

File: cmake/Modules/FindVJSON.cmake [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(
+    VJSON
+    DEFAULT_MSG
+    VJSON_LIBRARIES
+    VJSON_INCLUDE_DIRS
+)
+
+if(VJSON_FOUND)
+    mark_as_advanced(VJSON_LIBRARIES VJSON_INCLUDE_DIRS)
+endif()
```
Comment (jblomer) [inline cmake/Modules/FindVJSON.cmake]: Missing newline

File: externals/googletest/src/makeHook.sh [hunk 5]
```diff
@@ -2,3 +2,6 @@
 
 make clean
 make
+
+cp -rv include/gtest $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv lib*.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/googletest/src/makeHook.sh]: Missing newline

File: externals/leveldb/src/makeHook.sh [hunk 6]
```diff
@@ -3,3 +3,6 @@
 make clean
 CFLAGS="$CVMFS_BASE_C_FLAGS -fPIC" CXXFLAGS="$CVMFS_BASE_CXX_FLAGS -fPIC" make
 strip -S libleveldb.a
+
+cp -rv include/leveldb $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv libleveldb.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/leveldb/src/makeHook.sh]: Missing newline

File: externals/libcurl/src/makeHook.sh [hunk 7]
```diff
@@ -4,3 +4,4 @@ make clean
 make
 # Don't strip debug dymbols
 # strip -S lib/.libs/libcurl.a
+make install
```
Comment (jblomer) [inline externals/libcurl/src/makeHook.sh]: Missing newline

File: externals/pacparser/src/makeHook.sh [hunk 8]
```diff
@@ -24,3 +24,8 @@ echo "finished creating static link library for libpacparser"
 echo "stripping debug symbols for libpacparser..."
 strip -S libpacparser.a
 echo "finished building libpacparser.a"
+
+# Install
+cd ../../
+cp -v src/*.h $EXTERNALS_INSTALL_LOCATION/include/
+cp -v $static_result_dir/libpacparser.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/pacparser/src/makeHook.sh]: Missing newline

File: externals/sha2/src/makeHook.sh [hunk 9]
```diff
@@ -4,3 +4,5 @@ make clean
 make CVMFS_BASE_C_FLAGS="$CVMFS_BASE_C_FLAGS"
 strip -S libsha2.a
 
+cp -rv *.h $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv libsha2.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/sha2/src/makeHook.sh]: Missing newline

File: externals/sha3/src/makeHook.sh [hunk 10]
```diff
@@ -34,3 +34,5 @@ make clean
 make CVMFS_CASE_C_FLAGS="$CVMFS_BASE_C_FLAGS" ARCH=$(cat arch)
 strip -S libsha3.a
 
+cp -v *.h $EXTERNALS_INSTALL_LOCATION/include/
+cp -v libsha3.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/sha3/src/makeHook.sh]: Missing newline

File: externals/sqlite3/src/makeHook.sh [hunk 11]
```diff
@@ -3,3 +3,6 @@
 make clean
 make CVMFS_BASE_C_FLAGS="$CVMFS_BASE_C_FLAGS"
 strip -S libsqlite3.a
+
+cp -rv *.h $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv libsqlite3.a $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/sqlite3/src/makeHook.sh]: missing newline

File: externals/tbb/src/makeHook.sh [hunk 12]
```diff
@@ -9,6 +9,17 @@ fi
...
+# Install
+if [ x"$(uname)" = x"Darwin" ]; then
+  shlib_ext="dylib"
+else
+  shlib_ext="so"
+fi
+cp -rv include/serial $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv include/tbb $EXTERNALS_INSTALL_LOCATION/include/
+cp -rv build_debug/libtbb*.${shlib_ext}* $EXTERNALS_INSTALL_LOCATION/lib/
+cp -rv build_release/libtbb*.${shlib_ext}* $EXTERNALS_INSTALL_LOCATION/lib/
```
Comment (jblomer) [inline externals/tbb/src/makeHook.sh]: Missing newline

**Comment by radupopescu** [comment]
Hi, thanks for the feedback. I'll add the newlines, to avoid any potential warnings.

By the way, since we're using `find_package` across the board, in the CMake output we now have a report with the description of each linked library, so it's easier to track which one is linked statically vs dynamically.

**Pull request merged**